### PR TITLE
Print atlas features to PDF

### LINKED ISSUE #1714: Make "delete feature" accessible from the feature form
Currently features have to be deleted through a long press on the identify list.

The title bar should be cleaned up and possibly a ⋮menu be added with advanced functionality.

### PR DESCRIPTION
**Pull request by nirvn**
This PR implements a print atlas features to PDF function which enables a print button attached to focused/selected feature(s) on vector layers found to be atlas coverage layers in any of the print layouts for the currently loaded project:

The print action is available in a new feature form menu button when viewing the attributes:
image

The action is _also_ accessible via the feature list when the multi-selection mode is on, allowing for multiple atlas features to be exported at the same time:
image

As can be seen in the screenshot above, a revamp of the navigation bar had to be done to make space for this feature. This menu approach is forward thinking and gives us plenty of roam to add additional actions in the future without cluttering QField's interface. Instant bonus: the feature deletion mechanism is now much more visible.

Screen casts of the developed feature in action. First, printing of single atlas feature:
 
https://user-images.githubusercontent.com/1728657/109597940-d5d16b80-7b4a-11eb-9bd1-d18d66e46946.mp4  

Printing of multiple atlas features in a single operation via feature selection:

https://user-images.githubusercontent.com/1728657/109597959-dcf87980-7b4a-11eb-9d28-3d92c727f2c3.mp4 

As demonstrated in the screencasts, the feature implementation allows for more than one atlas-enabled layout to be linked to a given layer. If more than one atlas-enabled layout matches a given layer, a sub-menu will pop up for the user to select which atlas he/she wants to print. If only one match is found, the sub-menu is skipped and the atlas feature(s) is/are exported to PDF immediately.

Also showcased in the videos above is the use of QGIS' powerful symbology and expression engine to highlight the printed atlas feature(s) by changing the symbol style for the atlas feature. This is done via data-defined properties or rule-based renderer using a simple expression:  `$id = @atlas_featureid`.

Fixes #1714

### CODE REVIEW

File: src/core/qgismobileapp.cpp [hunk 1]
```diff
@@ -917,6 +927,56 @@ void QgisMobileapp::print( int layoutIndex )
...
+    return;
+
+  QStringList ids;
+  for( const auto id : featureIds )
+  {
+    ids << QString::number( id );
+  }
+
+  QString error;
+  layoutToPrint->atlas()->setFilterExpression( QStringLiteral( "$id IN (%1)" ).arg( ids.join( ',' ) ), error );
```
Comment (signedav) [inline src/core/qgismobileapp.cpp:953]: Would it be possible to keep the existing filter expression in the layout and add the feature filter with an "AND"?

Comment (nirvn) [reply src/core/qgismobileapp.cpp:953]: @signedav , I thought about it but I don't think it makes sense here. We _only_ want to print the selected feature(s) we're requesting.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e41c6ed8ed9690f114f3e55b7cf6ff6be81ce85d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1f9a962780326d884b4c1626247b8641da4ecc17-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1f9a962780326d884b4c1626247b8641da4ecc17-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1f9a962780326d884b4c1626247b8641da4ecc17-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1f9a962780326d884b4c1626247b8641da4ecc17-x86.apk)

File: src/core/qgismobileapp.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/qgismobileapp.cpp:953]: We also risk ending up with zero features. Say the filter has a generic expression like "MY_FIELD" != 'VALUE_1' and the user is requesting to print an atlas feature that is excluded by the pre-existing expression, that'd be confusing and  ultimately not what the user wants.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 6b973726c6eb4b2db6de840d68e1a39149ae0a0d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-29069a9e5cccefbebbee7c70c2ee5fc29868f76d-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-29069a9e5cccefbebbee7c70c2ee5fc29868f76d-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-29069a9e5cccefbebbee7c70c2ee5fc29868f76d-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-29069a9e5cccefbebbee7c70c2ee5fc29868f76d-x86.apk)

File: src/core/qgismobileapp.cpp (hunk 1 above)
Comment (signedav) [reply src/core/qgismobileapp.cpp:953]: Fair enough.

Another thing: While testing it did not work that it only printed the selected features. Like here (careful, it's not the same like in the clickup ticket) 
[GCK-Rto.zip](https://github.com/opengisch/QField/files/6062264/GCK-Rto.zip)

Comment (nirvn) [reply src/core/qgismobileapp.cpp:953]: @signedav , can you attach a screencast or steps to reproduce?

Comment (signedav) [reply src/core/qgismobileapp.cpp:953]: see #1732 - to double check I created a clean atlas (might be used later in the demo projects) - you can find there a screencast and reproduce it with these projects (advanced beefarming has single file and simple beefarming has one file per feature).

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 2df99abfeed3bffc2282ad6107511b43b9f99a55:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-099e9378ba0ee8d907f5d9002c861c132cb16071-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-099e9378ba0ee8d907f5d9002c861c132cb16071-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-099e9378ba0ee8d907f5d9002c861c132cb16071-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-099e9378ba0ee8d907f5d9002c861c132cb16071-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ba08c696f3ca3e11687771861090fa00d5a298ee:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-bda926777180d0b789ec3394e237beca9bba15bb-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-bda926777180d0b789ec3394e237beca9bba15bb-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-bda926777180d0b789ec3394e237beca9bba15bb-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-bda926777180d0b789ec3394e237beca9bba15bb-x86.apk)

**Comment by nirvn** [comment]
@signedav , all issues spotted during your testing session have been fixed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 3a572629facda4f11215ea4090d35311c12a853b:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-6a223ed888603156abcb092250d6b437e58b8be0-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-6a223ed888603156abcb092250d6b437e58b8be0-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-6a223ed888603156abcb092250d6b437e58b8be0-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-6a223ed888603156abcb092250d6b437e58b8be0-x86.apk)

**Comment by m-kuhn** [comment]
Nice 🎉 
And suddenly we need a task monitor where there is a spinner saying "printing pdf"

**Comment by nirvn** [comment]
@m-kuhn , that print export to PDF is done on the main thread and blocks the UI (same in QGIS), not sure there's a way ATM to provide a moving/spinning feedback.

**Comment by m-kuhn** [comment]
Upside: the toast stays visible.

**Comment by nirvn** [comment]
@m-kuhn , yeah, that's why I added the timer a while back, so at least it freezes on a frame that tells you something's happening :)

**Comment by nirvn** [comment]
@m-kuhn , one open question I'd like your opinion on: should we hide or disable the atlas print menu item when a given vector layer doesn't have atlas-driven layout(s) linked? I was 50-50 but ultimately settled on disabling instead of hiding because a/ it keeps ordering and spacing consistent (helps muscle memory), and b/ it helps "advertising" this feature to people who might then go "hmm, I wonder what that is, let's check the documentation... oh wow!".

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7b27f2824a8a3c1278295bb5fe88c84cf0c3fcaf:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-536803d9e8c4bdcd77dcd7ce3faf5529e47f61ac-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-536803d9e8c4bdcd77dcd7ce3faf5529e47f61ac-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-536803d9e8c4bdcd77dcd7ce3faf5529e47f61ac-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-536803d9e8c4bdcd77dcd7ce3faf5529e47f61ac-x86.apk)

**Comment by m-kuhn** [comment]
> @m-kuhn , one open question I'd like your opinion on: should we hide or disable the atlas print menu item when a given vector layer doesn't have atlas-driven layout(s) linked? I was 50-50 but ultimately settled on disabling instead of hiding because a/ it keeps ordering and spacing consistent (helps muscle memory), and b/ it helps "advertising" this feature to people who might then go "hmm, I wonder what that is, let's check the documentation... oh wow!".

Very good question. And I agree I'd leave it as you propose. We can still start kicking out disabled entries in the future if things become too crowded.

**Comment by signedav** [comment]
1. It works :tada:

**Comment by signedav** [comment]
The disabled button thing was one of the things I wanted to bring in. 
There is the case where the user is on the wrong layer - then I think it's good to have it disabled but visible - otherwise the user could think, the feature is gone. 
But another case is when there is no print layout in the project. So people that don't use printing there, and I'm more tending to not bother them with a disabled button. Good point with the "advertising", but on the other hand we always say we keep the UI as simple as possible.

In case we decide to hide the button sometimes (and break the static menu order) then I would hide the merging functionality on having only one feature selected.

 And don't you think deleting is a more used functionality than printing? Shouldn't it be on top? (no strong opinion)

**Comment by signedav** [comment]
And another small thing: On feature selection mode, the cancel-button is visible. When having a selection it changes to "not-selection" mode and on having no selection it closes the drawer (but keeps it for the next time in selection mode). I would prefer to make it either exiting the selection-mode or closing the drawer always (I think I prefer the first).

In case it will always exit the selection-mode, then I wonder if it would make sense to remove the "toggle feature selection" from the menu and just have a toggle instead of this cancel button always.

**Comment by signedav** [comment]
On exporting multiple PDF files (one per feature) it does not open them in the end. That's good because otherwise the user needs to close all of them. But there is no feedback neither if they are successfully exported and where it's stored? 
Maybe there could easily be a toast offered "Multiple documents exported to Documents/QField"?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 28a6d28cb0507b4cedca7e3bdf6ad45ae6689faf:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9053a62d8fa00ead812b80868bd5b87013e3eb26-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9053a62d8fa00ead812b80868bd5b87013e3eb26-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9053a62d8fa00ead812b80868bd5b87013e3eb26-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9053a62d8fa00ead812b80868bd5b87013e3eb26-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 1d2ccf3df10598a45b0e67418f7b654d669c4aff:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9bd3f2d4a1145f738c8747c2b8a9035e7ed456c9-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9bd3f2d4a1145f738c8747c2b8a9035e7ed456c9-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9bd3f2d4a1145f738c8747c2b8a9035e7ed456c9-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-9bd3f2d4a1145f738c8747c2b8a9035e7ed456c9-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for d15741b377efe20178a9c9c5fc57da7a1394cabd:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1ff1c3bdddfc1552190c289b8444b8008ce203d4-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1ff1c3bdddfc1552190c289b8444b8008ce203d4-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1ff1c3bdddfc1552190c289b8444b8008ce203d4-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-1ff1c3bdddfc1552190c289b8444b8008ce203d4-x86.apk)

**Comment by signedav** [comment]
Customer's happy but mentioned this:

> The only issue identified concerns the option "Predefined scale (best fit)" (Atlas function) which seems not working (the scale remains fixed).
The option « Margina round feature » and « Fixed scale » work.

I haven't tested myself. Maybe you have an idea @nirvn

**Comment by nirvn** [comment]
@signedav , good catch on the client's part, fixed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 9761a3fe204acfa7398390262a0b028cabc8040b:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-f01cf582d68af493e31019f0fef93ba6fd5a1a39-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-f01cf582d68af493e31019f0fef93ba6fd5a1a39-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-f01cf582d68af493e31019f0fef93ba6fd5a1a39-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-f01cf582d68af493e31019f0fef93ba6fd5a1a39-x86.apk)

File: src/core/qgismobileapp.cpp [hunk 2]
```diff
@@ -912,9 +923,90 @@ void QgisMobileapp::print( int layoutIndex )
...
+    QgsLayoutExporter exporter = QgsLayoutExporter( layoutToPrint );
+    QgsLayoutExporter::ExportResult result;
+    if ( layoutToPrint->customProperty( QStringLiteral( "singleFile" ), true ).toBool() )
+    {
+      result = exporter.exportToPdf( layoutToPrint->atlas(), destination, pdfSettings, error );
+      if ( result == QgsLayoutExporter::Success )
+        PlatformUtilities::instance()->open( destination );
+    } else {
+      result = exporter.exportToPdfs( layoutToPrint->atlas(), destination, pdfSettings, error );
+#ifndef Q_OS_ANDROID
```
Comment (signedav) [inline src/core/qgismobileapp.cpp:1001]: So it should open every single file on "not android"?

Comment (nirvn) [reply src/core/qgismobileapp.cpp:1001]: No, on windows/linux, it opens the container folder.

**Comment by signedav** [comment]
Regarding this:

**Comment by signedav** [comment]
And - just my opinion - but I would prefer a dynamic menu much more. Such things I dislike:
WhatsApp Image 2021-03-04 at 08 13 01

An idea: Maybe the three-dots menu could be a toggle-selection button and when selection is activated it turns to three-dots? 

But it does not need to be part of this PR...

**Comment by signedav** [comment]
But beside these UI complains - I really like this great feature. It's a awesome functionality and the code here looks solid. Great work!

File: src/core/qgismobileapp.cpp (hunk 2 above)
Comment (signedav) [reply src/core/qgismobileapp.cpp:1001]: Ah perfrct

**Comment by nirvn** [comment]
@signedav , could we merge this and address UI/UX tweaks in follow up PRs?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e924967f23daf877c6589ce114ef767584e32683:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-5fd22b9c8e787437d6e6bb558b401a6388757cb8-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-5fd22b9c8e787437d6e6bb558b401a6388757cb8-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-5fd22b9c8e787437d6e6bb558b401a6388757cb8-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1731-5fd22b9c8e787437d6e6bb558b401a6388757cb8-x86.apk)

**Review by signedav — APPROVED** [review]
:+1:

**Pull request merged**